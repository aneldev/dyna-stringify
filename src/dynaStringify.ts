export const dynaStringify = (
  value: any,
  options?: {
    spaces?: number;
    circularText?: string | ((value: any) => string);
  },
): string => {
  const {
    spaces = 0,
    circularText = '[CircularRef]',
  } = options || {};
  return JSON.stringify(value, getCircularReplacer(circularText), spaces);
};

const getCircularReplacer = (circularText: string | ((value: any) => string)) => {
  const scanned: any[] = [];
  return (key: string, value: any) => {
    if (typeof value === "object" && value !== null) {
      if (scanned.indexOf(value) > -1) {
        if (typeof circularText === "string") return circularText;
        return circularText(value);
      }
      scanned.push(value);
    }
    return value;
  };
};
