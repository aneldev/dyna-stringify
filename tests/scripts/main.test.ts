import "jest";

import {dynaStringify} from "../../src";

describe('dynaStringify', () => {
  it('Simple object, minimal config', () => {
    const value = {
      name: 'John',
      age: 32,
      cars: ['VW', 'Plymouth'],
    };
    expect(dynaStringify(value)).toMatchSnapshot();
  });

  it('Simple object, spaces 2', () => {
    expect(
      dynaStringify({
          name: 'John',
          age: 32,
          cars: ['VW', 'Plymouth'],
        },
        {
          spaces: 2,
        })
    )
      .toMatchSnapshot();
  });

  it('Simple object, with circular reference', () => {
    const value: any = {
      name: 'John',
      age: 32,
      cars: ['VW', 'Plymouth'],
    };
    value.value = value;
    expect(
      dynaStringify(
        value,
        {
          spaces: 2,
        })
    )
      .toMatchSnapshot();
  });

  it('Simple object, with custom circular reference text', () => {
    const value: any = {
      name: 'John',
      age: 32,
      cars: ['VW', 'Plymouth'],
    };
    value.value = value;
    expect(
      dynaStringify(
        value,
        {
          spaces: 2,
          circularText: '!REF',
        })
    )
      .toMatchSnapshot();
  });

  it('Simple object, with circular reference in array', () => {
    const value: any = {
      name: 'John',
      age: 32,
      cars: ['VW', 'Plymouth'],
    };
    value.cars.push(value);
    expect(
      dynaStringify(
        value,
        {
          spaces: 2,
        })
    )
      .toMatchSnapshot();
  });

  it('Simple object, with callback for circular reference text', () => {
    const value: any = {
      name: 'John',
      age: 32,
      cars: ['VW', 'Plymouth'],
    };
    value.cars.push(value);
    value.cars.push(value.cars);
    expect(
      dynaStringify(
        value,
        {
          spaces: 2,
          circularText: (value: any) => {
            if (value.name) return `Circular reference for name: ${value.name}`;
            return '!REF';
          },
        })
    )
      .toMatchSnapshot();
  });
});
