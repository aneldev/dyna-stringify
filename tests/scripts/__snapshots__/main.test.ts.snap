// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynaStringify Simple object, minimal config 1`] = `"{\\"name\\":\\"John\\",\\"age\\":32,\\"cars\\":[\\"VW\\",\\"Plymouth\\"]}"`;

exports[`dynaStringify Simple object, spaces 2 1`] = `
"{
  \\"name\\": \\"John\\",
  \\"age\\": 32,
  \\"cars\\": [
    \\"VW\\",
    \\"Plymouth\\"
  ]
}"
`;

exports[`dynaStringify Simple object, with callback for circular reference text 1`] = `
"{
  \\"name\\": \\"John\\",
  \\"age\\": 32,
  \\"cars\\": [
    \\"VW\\",
    \\"Plymouth\\",
    \\"Circular reference for name: John\\",
    \\"!REF\\"
  ]
}"
`;

exports[`dynaStringify Simple object, with circular reference 1`] = `
"{
  \\"name\\": \\"John\\",
  \\"age\\": 32,
  \\"cars\\": [
    \\"VW\\",
    \\"Plymouth\\"
  ],
  \\"value\\": \\"[CircularRef]\\"
}"
`;

exports[`dynaStringify Simple object, with circular reference in array 1`] = `
"{
  \\"name\\": \\"John\\",
  \\"age\\": 32,
  \\"cars\\": [
    \\"VW\\",
    \\"Plymouth\\",
    \\"[CircularRef]\\"
  ]
}"
`;

exports[`dynaStringify Simple object, with custom circular reference text 1`] = `
"{
  \\"name\\": \\"John\\",
  \\"age\\": 32,
  \\"cars\\": [
    \\"VW\\",
    \\"Plymouth\\"
  ],
  \\"value\\": \\"!REF\\"
}"
`;
